### PR TITLE
[Refactor] Don't publicly expose `ZMUser` in Typing Classes

### PR DESCRIPTION
**Pull request by johnxnguyen**
## What's new in this PR?

This PR is part of an ongoing effort to all public references that leak concrete user objects. See the architecture issues for more info: https://github.com/wireapp/ios-architecture/issues/89

Public exposure of `ZMUser` has been removed from the following cluster of files:

- `ZMTyping.m`
- `ZMTypingUsers.m`
- `ZMTypingUsersTimeout.m`

`ZMUser` was publicly exposed due to Objc/Swift interoperability. By converting these classes to Swift, I was able to make them all internal.

### Note
I also converted the tests to Swift too.

### CODE REVIEW

File: Source/Data Model/Typing.swift [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see http://www.gnu.org/licenses/.
+//
+
+import Foundation
+
+class Typing {
+
+    #if DEBUG
+    public static var defaultTimeout: TimeInterval = 60
```
Comment (typfel) [inline Source/Data Model/Typing.swift]: Debug `defaultTimeout` is the same as non-debug.

Comment (johnxnguyen) [reply Source/Data Model/Typing.swift]: The difference is that in DEBUG it is mutable (for testing purposes).

Comment (typfel) [reply Source/Data Model/Typing.swift]: Ah 🧐

File: Source/Data Model/TypingUsers.swift [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+        }
+    }
+}
+
+@objc public extension ZMConversation {
+
+    @objc (setIsTyping:)
+    func setIsTyping(_ isTyping: Bool) {
+        TypingStrategy.notifyTranscoderThatUser(isTyping: isTyping, in: self)
+    }
```
Comment (billypchan) [inline Source/Data Model/TypingUsers.swift]: Can we change this to a getter `var isTyping: Bool`?

File: Source/Data Model/Typing.swift [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+    }
+
+}
+
+// MARK: - TimerClient
+
+extension Typing: ZMTimerClient {
+
+    func timerDidFire(_ timer: ZMTimer!) {
+        guard timer == expirationTimer else { return }
```
Comment (typfel) [inline Source/Data Model/Typing.swift]: Maybe do a `===` for pointer comparison which is what `==` means in Objective-C

File: Source/Data Model/TypingUsers.swift (hunk 2 above)
Comment (johnxnguyen) [reply Source/Data Model/TypingUsers.swift]: Oh wait, this is not a stored property though.

File: Source/Data Model/Typing.swift (hunk 3 above)
Comment (johnxnguyen) [reply Source/Data Model/Typing.swift]: Done

File: Source/Data Model/TypingUsersTimeout.swift [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+        let userIds = timeouts.keys
+            .filter { $0.conversationObjectId == conversation.objectID }
+            .map(\.userObjectId)
+
+        return Set(userIds)
+    }
+
+    func pruneConversationsThatHaveTimoutAfter(date pruneDate: Date) -> Set<NSManagedObjectID> {
+        let keysToRemove = timeouts
+            .filter { $0.value < pruneDate }
```
Comment (johnxnguyen) [inline Source/Data Model/TypingUsersTimeout.swift]: @typfel I was wondering about this. The method says "prune conversations that have timeout **after** this date", but we remove entries whose date is actually **before** the prune date. This was a direct translation from ObjC, and the tests are fine. So I'm guess it's just the wrong name for the method?

Comment (typfel) [reply Source/Data Model/TypingUsersTimeout.swift]: Cleared in person.

File: Source/Data Model/Conversation+TypingUsers.swift [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+        TypingStrategy.notifyTranscoderThatUser(isTyping: isTyping, in: self)
+    }
+
+    var typingUsers: [UserType] {
+        guard let users = managedObjectContext?.typingUsers?.typingUsers(in: self) else { return [] }
+        return Array(users)
+    }
+
+    /// Strictly for UI tests. Remove once mockable conversation abstraction exists.
+    func setTypingUsers(_ users: [UserType]) {
```
Comment (johnxnguyen) [inline Source/Data Model/Conversation+TypingUsers.swift]: @typfel , regarding the UI tests that need to set the typing users for the conversation: I couldn't find a way to mock the typing users without a large scale refactoring. So I've decided to make this public method for testing purposes.

Ideally we shouldn't expose this, but also ideally we should have a conversation abstraction like `UserType` with the aim of removing all CoreData from our UI tests. I'll create a ticket to track this issue.

Comment (typfel) [reply Source/Data Model/Conversation+TypingUsers.swift]: 👍

**Pull request merged**